src/main/lock.ts

```diff
@@ -0,0 +1,44 @@
+// ==LICENSE-BEGIN==
+// Copyright 2017 European Digital Reading Lab. All rights reserved.
+// Licensed to the Readium Foundation under one or more contributor license agreements.
+// Use of this source code is governed by a BSD-style license
+// that can be found in the LICENSE file exposed on Github (readium) in the project repository.
+// ==LICENSE-END==
+
+import * as debug_ from "debug";
+import { app } from "electron";
+import { container } from "readium-desktop/main/di";
+import { WinRegistry } from "readium-desktop/main/services/win-registry";
+
+import { cli } from "./cli/process";
+
+// Logger
+const debug = debug_("readium-desktop:main:lock");
+
+export function lockInstance() {
+    const gotTheLock = app.requestSingleInstanceLock();
```

Comment:
Any attempt to launch the Thorium app more than once must result in the process being "terminated". In other words, an app that does not obtain the "Single Instance Lock" must call `app.quit()` **immediately**. When that happens, the first / original application process receives the `"second-instance"` event (or, in the case of MacOS, the `"open-file" event`).

So, I think we should have code like:
```
const gotTheLock = app.requestSingleInstanceLock();
if (!gotTheLock) {
    app.quit();
} else {
    ...
}

```

Reply:
@panaC any comments on the "single instance" design pattern? Is there anything blocking or unclear about it?

Electron documentation:

`app.requestSingleInstanceLock()`:
https://github.com/electron/electron/blob/master/docs/api/app.md#apprequestsingleinstancelock

`second-instance`:
https://github.com/electron/electron/blob/master/docs/api/app.md#event-second-instance

Reply:
thank you Daniel,

It's seems good to me

I have tested on mac and win, both of this platforms work

I showed to Laurent: he is agree with that.

I don't understand, why you blocking this PR.

Reply:
Maybe I am missing something obvious, but I hope my code review is correct :)

I already explained the rationale in my previous comment, let me try again:

The Electron design pattern dictates that `app.quit()` (or `app.exit()`) must be called immediately when the "single instance lock" is not obtained, because Electron gives control to the existing/original instance via the `second-instance` event (or `open-file`, in a mutually-exclusive manner).

In the current PR, the code after the call to `const gotTheLock = lockInstance()` continues to execute, initializes Yargs, and inside the Yargs handlers the boolean condition `!gotTheLock` is checked which is nonsensical because an Electron app that does not have a self instance lock  should have aborted immediately.

The solution is quite simple (copy-paste from my original comment above):
```
const gotTheLock = app.requestSingleInstanceLock();
if (!gotTheLock) {
    app.quit();
} else {
    ...
}
```

Reply:
When the code is fixed as per the Electron's documentation, please also remove unnecessary code such as:
```
            // if it's the main instance
            if (gotTheLock) {
```

or:

```
            // if the app is not started or it's the second instance
            if (!app.isReady || !gotTheLock) {
```

etc.

(there is no such thing as a running app that has been refused the "single instance lock", as by definition it must be terminated immediately)